File: Top Interview Questions/L160. 相交链表.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

//  function getIntersectionNode(headA: ListNode | null, headB: ListNode | null): ListNode | null {
//     let set= new Set()
//     while(headA!=null){
//         set.add(headA)
//         headA=headA.next
//     }
//     while(headB!=null){
//         if(set.has(headB)){
//             return headB
//         }
//         headB=headB.next
//     }
//     return null
// };

function getIntersectionNode(headA: ListNode | null, headB: ListNode | null): ListNode | null {
        if (headA === null || headB === null) {
            return null;
        }
        let pA = headA, pB = headB;
        while (pA !== pB) {
            pA = pA === null ? headB : pA.next;
            pB = pB === null ? headA : pB.next;
        }
        return pA;
};

File: java-devildyw/src/main/java/top/dyw/leetcode/Hot100/L160相交链表.java
package top.dyw.leetcode.Hot100;

/*
 * @lc app=leetcode.cn id=160 lang=java
 *
 * [160] 相交链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class L160相交链表 {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        // 走完自己的再走对方的
        ListNode p1 = headA, p2 = headB;
        while (p1!=p2) {
            p1 = p1 != null ? p1.next : headB;
            p2 = p2 != null ? p2.next : headA;
        }

        return p1;
    }
}
// @lc code=end
